### ISSUE TITLE
socket-connect for CCL causing subsequent "Bad File Descriptor during Read" on large files

### ISSUE THREAD

**Issue by gendl** [usocket/usocket#103]
Hi,

 In upgrading from 0.8.3 to 0.8.6 I start to get errors on CCL in my test suite where it happens to do several http client calls (with zaserve client or drakma) then open and try to read a large file immediately. 
 
 The problem is isolated to function `socket-connect` from `backend/clozure.lisp`. 
 
 You can replicate as follows:
 
 
 ```
 (defun replicate ()
  (make-web-request) (make-web-request)(make-web-request)
  ;; (sleep 1)  ;; if this sleep it put here the error does not happen.
  (read-large-file))

(defun make-web-request ()
  (drakma:http-request "http://google.com"))

(defun read-large-file ()
  (with-open-file (in <put a large file pathname here>)
    (read in)))
 ```
 
 Then call 
 
 ```
 (replicate)
 ```
 
 You may well get the "Bad File Descriptor during Read"
 
 I think it may be leaving a dangling openmcl socket when it tries to call `openmcl-socket:make-socket` and fails when trying with ipv6 (as it now does explicitly all the time). 
 
 As noted in the code above, the error only happens if a large file read is attempted immediately after the http client call.  
 
 I wonder if there is something the CCL socket code is not cleaning up, when a call to `openmcl-socket:make-socket` is tried and fails with an error. Or does it have to do with the `:deadline` `:nodelay` or `:connect-timeout` 
 
 Anyway i can look into this further but it's late here now so I thought i'd post the simple example for replicating, in case anyone cares to try replicating. 
 
 Thanks,
 
  Dave

**Comment by gendl** [comment]
Note that to replicate successfully, you probably need to not have an active ipv6 net connection. The resolver for e.g. google.com will return at least one ipv6 address, but then if you try to open an active socket on that address but you don't have a net interface to support ipv6, the make-socket will fail and that's when you can replicate the bug.

**Comment by easye** [comment]
@gendl I can't seem to replicate, but not sure about two points:

1.  What is a large file?  Is 1Mib of bytes enough?

2.  I don't seem to get a "bad file descriptor during read", but rather I get a 

```
Unexpected end of file on #<BASIC-FILE-CHARACTER-INPUT-STREAM ("/home/mevenson/tmp/xx"/\
4 UTF-8) #x30200233F7DD>, near position 1048576  
```

which for a 1Mib files of zeros created via `dd if=/dev/zero of=~/tmp/xx bs=1M count=1 `, this makes sense as an error.

**Comment by gendl** [comment]
@easye [Here](https://gitlab.common-lisp.net/gendl/gendl/-/blob/release/1598/regression/data/huge-boxes-sequence.data) is where you can download the actual <large file> I've been testing with.

Did you disable ipv6 support in your networking before trying to replicate?

**Comment by easye** [comment]
> @easye [Here](https://gitlab.common-lisp.net/gendl/gendl/-/blob/release/1598/regression/data/huge-boxes-sequence.data) is where you can download the actual file I've been testing with.

A more correct link is via 

```wget https://gitlab.common-lisp.net/gendl/gendl/-/raw/release/1598/regression/data/huge-boxes-sequence.data?inline=false -O huge-boxes-sequence.data``` to a file 

```
https://gitlab.common-lisp.net/gendl/gendl/-/raw/release/1598/regression/data/huge-boxes-sequence.data?inline=false
```
> 
> Did you disable ipv6 support in your networking before trying to replicate?

**Comment by easye** [comment]
> Did you disable ipv6 support in your networking before trying to replicate?

No.  This has to be done at the link level, or is removing something from ANSI `**features**` enough?

**Comment by gendl** [comment]
ipv6 would have to be disabled at the link level. It needs to be present in `*features*` but not working at the network level so that the ipv6 connection will be attempted and will fail, thus triggering the bug. What it's meant to do in that case is "fall back" to ipv4, which it does - and we want to retain that behavior. What we need to fix is the apparent side-effect from attempting the ipv6 connection, which appears to result in the "Bad File Descriptor" error downstream.

**Comment by gendl** [comment]
To be more precise - ipv6 doesn't necessarily have to be disabled globally in your whole OS - it's enough that the particular address you are doing the drakma call to meets two conditions:  1. It has an active AAAA record such that a dns lookup with `:address-family` `:internet6` will return an ipv6 address, and 2. Trying to connect to said ipv6 address fails (with a "no route to host" or similar, but I'm not sure whether the reason for failure matters).

**Comment by gendl** [comment]
I'm having trouble replicating this myself now - trying on my home iMac as well as Docker linux container on same. So far the only place I can replicate is running CCL in WSL Ubuntu on my LG Gram.. So please don't pull out your hair trying to replicate this..

**Comment by easye** [comment]
> So please don't pull out your hair trying to replicate this..

Standing down from trying to replicate.  @gendl find me on IRC to chat about next steps?

**Comment by easye** [comment]
> I am now replicating reliably on any Linux CCL 1.12 I try, including WSL Ubuntu and Docker.

I cannot replicate under a Linux vultr host.  I'll triage the exact details, but once I do, I will chase @gendl down on IRC.

**Comment by gendl** [comment]
I have established a quick way to predict whether this error will replicate on your platform. Basically the error will happen if your host or network is not configured for ipv6, yet the resolver returns ipv6 addresses if you ask explicitly for them. Here is how to determine if this is the case on your testing platform:

```
(setq remote (openmcl-socket:resolve-address :host "google.com"
                                                :port 80
                                                :socket-type :stream
                                                :address-family :internet6))
```

Assuming the above returns an ipv6 address, now do:

```
(openmcl-socket:make-socket
                 :type :stream
                 :address-family :internet6
                 :remote-address remote)
                 
```

If that returns a socket without errors, then your setup is configured for ipv6 and has ipv6 connectivity. If not, then not, and likely you will be able to replicate the error in this Issue description.

**Comment by gendl** [comment]
Note that if you have cloned the gendl repository, you can also just run the pipeline tests directly from a shell with 

```

cd gendl
./docker/pipeline-tests

```

**Comment by easye** [comment]
I have (finally) managed to replicate the "Bad File Descriptor during Read" issue on native Linux without reference to gendl.  

I am now rummaging through the SLIME inspector at the point where the condition is signalled after having read 491520 of 5259205 bytes.  More information when I have it…

**Comment by gendl** [comment]
@easye could you describe how you managed to replicate? Maybe @binghe could have a go at replicating (I only presume he might want to try that since he assigned himself here)?  It might be telling that doing e.g. a `(sleep 0.5)` after the http client call and before the big file read, makes it so the problem doesn't show up...

**Comment by easye** [comment]
I essentially followed your instructions on a cloud hosted Linux instance.I have a git repo with code and notes in an org file that may be of use.Away from a proper keyboard at the moment, so sharing the repo isn’t possible but I will do if folks are interested.Sent from my iPadOn May 18, 2023, at 20:48, Forked third-party libs ***@***.***> wrote:﻿
@easye could you describe how you managed to replicate? Maybe @binghe could have a go at replicating (I only presume he might want to try that since he assigned himself here)?  It might be telling that doing e.g. a (sleep 0.5) after the http client call and before the big file read, makes it so the problem doesn't show up...
